### PR TITLE
fix reactor#3539 takeUntil Predicate test before emit

### PR DESCRIPTION
**Pull request by AramMessdaghi9001**
This fixes issue #3539 
The documentation does not specify when takeUntil tests the predicate.
The current implementation does this after the element has been emitted. If the emitted element is modified by downstream consumers, this affects the predicate function.
This behavior might be confusing and lead to unwanted results.
This PR moves the test up and documents the behavior in order to avoid this caveat.

### CODE REVIEW

**Review by OlegDokuka — APPROVED** [review]

File: reactor-core/src/test/java/reactor/core/publisher/FluxTakeUntilPredicateTest.java [hunk 1]
```diff
@@ -160,8 +176,7 @@ public void predicateThrows() {
 			throw new RuntimeException("forced failure");
 		}).subscribe(ts);
 
-		ts.assertValues(1)
-		  .assertNotComplete()
+		ts.assertNotComplete()
```
Comment (OlegDokuka) [inline reactor-core/src/test/java/reactor/core/publisher/FluxTakeUntilPredicateTest.java:179]: seems to be unrelated change

**Review by OlegDokuka — CHANGES_REQUESTED** [review]

File: reactor-core/src/test/java/reactor/core/publisher/FluxTakeUntilPredicateTest.java (hunk 1 above)
Comment (AramMessdaghi9001) [reply reactor-core/src/test/java/reactor/core/publisher/FluxTakeUntilPredicateTest.java:179]: Actually the exception in the predicate here now happens before the element "1" is emitted, so "1" never gets out.

Comment (chemicL) [reply reactor-core/src/test/java/reactor/core/publisher/FluxTakeUntilPredicateTest.java:179]: I have doubts here. I understand the reasoning, but this is a potentially breaking change if anyone relied on this behaviour. The behaviour is not backwards compatible with regards to error handling. Assuming there is code out there that actually deals with the delivery of the item that then causes an exception when tested, we should preserve this behaviour, or address this change for 3.6 with a clear explanation this is breaking.

Comment (AramMessdaghi9001) [reply reactor-core/src/test/java/reactor/core/publisher/FluxTakeUntilPredicateTest.java:179]: The later one would be cleaner, wouldn't it? Shall I rebase? Whats the procedure on breaking changes?

Comment (OlegDokuka) [reply reactor-core/src/test/java/reactor/core/publisher/FluxTakeUntilPredicateTest.java:179]: I don't think there is breaking change unless there is a bug in the given test, even though predicate result is false value is delivered since return statement happens after the a.onNext. I will have a closer look to figure out what is wrong with the test

Comment (OlegDokuka) [reply reactor-core/src/test/java/reactor/core/publisher/FluxTakeUntilPredicateTest.java:179]: the onNext method should look like the following.

```java
@Override
		public void onNext(T t) {
			if (done) {
				Operators.onNextDropped(t, actual.currentContext());
				return;
			}

			boolean b;

			try {
				b = predicate.test(t);
			} catch (Throwable e) {
				// ensures value is delivered regardless the failure
				actual.onNext(t);
				onError(Operators.onOperatorError(s, e, t, actual.currentContext()));
				return;
			}

			actual.onNext(t);

			if (b) {
				s.cancel();

				onComplete();
			}
		}
```

having it like that will make the test pass. We have to preserve original behaviour

**Review by OlegDokuka — APPROVED** [review]

**Review by chemicL — APPROVED** [review]

**Comment by OlegDokuka** [comment]
@AramMessdaghi9001 thank you for your contribution! Keep it up!
